src/util/SimpleTimer.h

```diff
@@ -0,0 +1,192 @@
+    bool mActive;
+    std::chrono::steady_clock::time_point mStart;
+
+  public:
+    SimpleTimerContext(SimpleTimer<Duration>& timer);
+    ~SimpleTimerContext();
+    void Reset();
+    std::chrono::nanoseconds Stop();
+};
+
+template <typename Duration>
+SimpleTimer<Duration>::SimpleTimer(
+    medida::MetricsRegistry& registry, std::string const& domain,
+    std::string const& type, std::string const& base_name,
+    std::optional<std::chrono::nanoseconds> windowSize)
+    : mSum{registry.NewCounter({domain, type, std::string{base_name} + "sum"})}
+    , mCount{registry.NewCounter(
+          {domain, type, std::string{base_name} + "count"})}
+    , mMaxCounter{registry.NewCounter(
+          {domain, type, std::string{base_name} + "max"})}
+    , mMax{0}
+    , mLastUpdate{std::chrono::steady_clock::now()}
+    , mWindowSize{windowSize}
+{
+}
+
+template <typename Duration>
+SimpleTimer<Duration>::SimpleTimer(SimpleTimer<Duration>&& other)
+    : mSum{other.mSum}
+    , mCount{other.mCount}
+    , mMaxCounter{other.mMaxCounter}
+    , mMax{other.mMax}
+    , mLastUpdate{other.mLastUpdate}
+    , mWindowSize{other.mWindowSize} {};
+
+template <typename Duration>
+void
+SimpleTimer<Duration>::syncMax()
+{
+
+    MutexLocker lock{mLock};
+    syncMaxUnlocked();
+}
+
+template <typename Duration>
+void
+SimpleTimer<Duration>::syncMaxUnlocked()
+{
+
+    mMaxCounter.set_count(mMax);
+    mMax = 0;
+    mLastUpdate = std::chrono::steady_clock::now();
+}
+
+template <typename Duration>
+std::int64_t
+SimpleTimer<Duration>::count() const
+{
+    return mCount.count();
+}
+
+template <typename Duration>
+void
+SimpleTimer<Duration>::Update(std::chrono::nanoseconds d)
+{
+    auto converted = std::chrono::duration_cast<Duration>(d).count();
+    mSum.inc(converted);
+    mCount.inc(1);
+    {
+        MutexLocker lock{mLock};
+        mMax = std::max(mMax, converted);
+        if (mWindowSize.has_value() && std::chrono::steady_clock::now() >=
+                                           mWindowSize.value() + mLastUpdate)
+        {
+            syncMaxUnlocked();
+        }
+    }
+}
+
+template <typename Duration>
+SimpleTimerContext<Duration>
+SimpleTimer<Duration>::TimeScope()
+{
+    return SimpleTimerContext<Duration>(*this);
+}
+
+template <typename Duration>
+SimpleTimerContext<Duration>::SimpleTimerContext(SimpleTimer<Duration>& timer)
+    : mTimer(timer)
+{
+    Reset();
+};
+
+template <typename Duration> SimpleTimerContext<Duration>::~SimpleTimerContext()
+{
+    Stop();
+};
+
+template <typename Duration>
+void
+SimpleTimerContext<Duration>::Reset()
+{
+    mStart = std::chrono::steady_clock::now();
+    mActive = true;
+}
+
+template <typename Duration>
+std::chrono::nanoseconds
+SimpleTimerContext<Duration>::Stop()
+{
+    if (mActive)
+    {
+        auto dur = std::chrono::steady_clock::now() - mStart;
+        mTimer.Update(dur);
+        mActive = false;
+        return dur;
+    }
+    return Duration{0};
```

Comment:
Rather than returning `0` in this case, perhaps this should return `nullopt`. That allows callers to more easily detect if they made a mistake in reusing a stopped context without resetting it first.